Location Reassignment Phase 3 Fixes

mkangia opened the pull request:
Some issues found during first round of preliminary testing.

##### SUMMARY
1. A correction to logic for fetching household child cases
Depending on how cases are structured its possible to get the same case more than once so fetch unique cases
2. Load using objects to be able to load archived locations as well


##### FEATURE FLAG
`LOCATION_REASSIGNMENT`

mkangia committed:
get unique users

mkangia committed:
allow fetching archived locations

orangejenny commented:
Should there be test cases for either of these issues?

mkangia commented:
Agree that the second case would have been caught in tests if there were end to end tests for full process but the first one is more like a HQ thing on how cases are structured which should have been considered in the logic already.

mkangia commented on custom/icds/location_reassignment/utils.py:
```diff
@@ -52,21 +52,26 @@ def get_household_and_child_case_ids_by_owner(domain, household_case_id, owner_i
     parent_case_ids = [household_case_id]
     while parent_case_ids:
-        child_cases = get_child_cases(parent_case_ids)
+        child_cases = [case for case in get_child_cases(parent_case_ids, iterated_case_ids)]
```
updated [later](https://github.com/dimagi/commcare-hq/pull/27153/commits/e81fcda4d951aef3e8b03051937086b15e0575fa) to avoid the redundant iteration.

orangejenny commented on custom/icds/location_reassignment/utils.py:
```diff
@@ -52,21 +52,26 @@ def get_household_and_child_case_ids_by_owner(domain, household_case_id, owner_i
-                if case.owner_id == owner_id]
+                if case.case_id not in exclude_ids and case.owner_id == owner_id]
 
+    # keep a list of case ids already iterated to avoid duplicates
```
Given that `cases` is already a set, I don't quite follow why this is needed, won't duplicates already be eliminated?

esoergel commented on custom/icds/location_reassignment/utils.py:
```diff
@@ -52,28 +52,31 @@ def get_household_and_child_case_ids_by_owner(domain, household_case_id, owner_i
+            child_case_ids = [case.case_id for case in child_cases]
+            iterated_case_ids.update(child_case_ids)
+            cases.update([case for case in child_cases if case.type in case_types])
+            parent_case_ids = child_case_ids
```
As an alternative to passing `iterated_case_ids` to `get_child_cases`, you could subtract them from `child_case_ids` here:
```python
parent_case_ids = child_case_ids - iterated_case_ids
```
(it would have to be defined as a set for this to work, of course, but that shouldn't be a problem)

esoergel replied:
Oh, my approach prevents you from doing the lookup multiple times with a single ID, but it doesn't deduplicate the lookup results, as this approach does.  For that you'd have to deduplicate before updating `cases`

mkangia replied:
`iterated_case_ids` have already been excluded so they wont be present in child_case_ids to that set subtraction wont be useful?

esoergel replied:
They're excluded by passing it in to `get_child_cases` and skipping them there - I was saying you could instead do a simple set subtraction here, but that won't work with the current state, since `cases` is no longer a data structure which excludes duplicates.

mkangia committed:
Revert "get unique users"

This reverts commit 60ed78c4a79a0ea8bafaf0e2e8fe81be81f0e9ec.

mkangia committed:
get unique users

mkangia commented:
@orangejenny @esoergel 
this is now final and good to review.

mkangia committed:
update signature for optional argument

mkangia committed:
update url

esoergel commented on custom/icds/location_reassignment/utils.py:
```diff
@@ -51,29 +51,34 @@ def get_household_and_child_case_ids_by_owner(domain, household_case_id, owner_i
         return [case for case in
                 CaseAccessorSQL.get_reverse_indexed_cases(domain, ids)
-                if case.owner_id == owner_id]
+                if case.case_id not in exclude_ids and case.owner_id == owner_id]
```
Can intermediate cases be owned by someone other than `owner_id`?  Should the `owner_id` check instead happen when the cases are added to `cases`?

mkangia replied:
> Can intermediate cases be owned by someone other than owner_id

Ideally in case of ICDS-CAS all cases should be owned by the same user.

Good point, https://github.com/dimagi/commcare-hq/pull/27153/commits/6c6374317a9238e81ab86a6a351327713bc0c0b2

mkangia committed:
delay filtering out cases owned by owner

its possible that a case not owned by the owner has a child case which is yet to be iterated and is owned by the owner, so filter by owner only by the end
